dotnet/src/SemanticKernel.Skills/Skills.Memory.CosmosDB/CosmosMemoryStore.cs

```diff
@@ -0,0 +1,207 @@
+
+                    var rec = MemoryRecord.FromJson(
+                        record.MetadataString,
+                        new Embedding<float>(embeddingHost.Embedding.vector));
+
+                    return DataEntry.Create<IEmbeddingWithMetadata<TEmbedding>>(
+                                rec.Metadata.Id,
+                                (IEmbeddingWithMetadata<TEmbedding>)rec,
+                                record.Timestamp);
+                }
+            }
+        }
+
+        return null;
+    }
+
+    public async IAsyncEnumerable<string> GetCollectionsAsync([EnumeratorCancellation] CancellationToken cancel = default)
+    {
+        var container = this._client.GetContainer(this._databaseName, this._containerName);
+        var query = new QueryDefinition($"SELECT DISTINCT c.collectionId FROM c");
+        var iterator = container.GetItemQueryIterator<CosmosMemoryRecord>(query);
+
+        var items = await iterator.ReadNextAsync(cancel).ConfigureAwait(false);
+
+        foreach (var item in items)
+        {
+            yield return item.CollectionId;
+        }
+    }
+
+    public IAsyncEnumerable<(IEmbeddingWithMetadata<TEmbedding>, double)> GetNearestMatchesAsync(string collection, Embedding<TEmbedding> embedding, int limit = 1, double minRelevanceScore = 0)
+    {
+        if (limit <= 0)
+        {
+            return AsyncEnumerable.Empty<(IEmbeddingWithMetadata<TEmbedding>, double)>();
+        }
+
+        var asyncEmbeddingCollection = this.TryGetCollectionAsync(collection);
+        var embeddingCollection = asyncEmbeddingCollection.ToEnumerable().ToArray();
+
+        if (embeddingCollection == null || !embeddingCollection.Any())
+        {
+            return AsyncEnumerable.Empty<(IEmbeddingWithMetadata<TEmbedding>, double)>();
+        }
+
+        EmbeddingReadOnlySpan<TEmbedding> embeddingSpan = new(embedding.AsReadOnlySpan());
+
+        TopNCollection<IEmbeddingWithMetadata<TEmbedding>> embeddings = new(limit);
+
+        foreach (var item in embeddingCollection)
+        {
+            if (item.Value != null)
+            {
+                EmbeddingReadOnlySpan<TEmbedding> itemSpan = new(item.Value.Embedding.AsReadOnlySpan());
+                double similarity = embeddingSpan.CosineSimilarity(itemSpan);
+                if (similarity >= minRelevanceScore)
+                {
+                    embeddings.Add(new(item.Value, similarity));
+                }
+            }
+        }
+
+        embeddings.SortByScore();
+
+        return embeddings.Select(x => (x.Value, x.Score.Value)).ToAsyncEnumerable();
+    }
+
+    protected async IAsyncEnumerable<DataEntry<IEmbeddingWithMetadata<TEmbedding>>> TryGetCollectionAsync(string collectionName, [EnumeratorCancellation] CancellationToken cancel = default)
+    {
+        var container = this._client.GetContainer(this._databaseName, this._containerName);
+        var query = new QueryDefinition($"SELECT * FROM c WHERE c.collectionId = '{collectionName}'");
```

Comment:
I very much recommend using parameterized queries. https://learn.microsoft.com/en-us/dotnet/api/microsoft.azure.cosmos.querydefinition.withparameter?view=azure-dotnet#examples They are *much* more secure.

For example 
```
QueryDefinition query = new QueryDefinition(
    "SELECT * FROM c WHERE c.collectionId = @collectionName")
    .WithParameter("@collectionName", collectionName);
```